### PR TITLE
`AggressiveOptimization` in `InProcess` toolchains

### LINKED ISSUE #1934: Ensure WorkloadActionUnroll and similar are optimized if possible
If `{Workload|Overhead}Action{No}Unroll` are not optimized, and the benchmark method allocates a lot, the Tier0 versions of these methods may retain much more heap data than the optimized versions. This can completely change benchmark results.

Current behavior for CLR is that these methods are always eagerly optimized, because they all contain loops and we either do not have tiered compilation (full framework, pre-3.0 core) or have `COMPlus_TC_QuickJitForLoops=1` (core 3.0 - core 6.0). 

For .NET 7 we plan to change the default to `COMPlus_TC_QuickJitForLoops=0` and this leads to these methods not being optimized and significant behavior changes in roughly 50 or so benchmarks on github.com/dotnet/performance.

To preserve current behavior, these methods should always be optimized when running against .NET Core versions that can change tiering strategy. This can be done by adding `[MethodImpl(MethodImplOptions.AggressiveOptimziation)]` which is available in .Net Core 3.0 and up.

Alternatively one could consider restructuring things so that the benchmark delegate is always called from a method that doesn't have loops. This seems more disruptive.

See https://github.com/dotnet/performance/issues/2214#issuecomment-1052585650 and following comments for context.

cc @adamsitnik

### PR DESCRIPTION
**Pull request by timcassell**
Follow-up to #1935, fixes #1934 for in-process toolchains.

### CODE REVIEW

File: src/BenchmarkDotNet/Toolchains/InProcess/NoEmit/BenchmarkActionFactory_Implementations.cs [hunk 1]
```diff
@@ -14,6 +15,9 @@ namespace BenchmarkDotNet.Toolchains.InProcess.NoEmit
     /// <summary>Helper class that creates <see cref="BenchmarkAction"/> instances. </summary>
     public static partial class BenchmarkActionFactory
     {
+        // Not available in netstandard2.0, so just use the value casted to enum.
+        public const MethodImplOptions AggressiveOptimization = (MethodImplOptions) 512;
```
Comment (AndreyAkinshin) [inline src/BenchmarkDotNet/Toolchains/InProcess/NoEmit/BenchmarkActionFactory_Implementations.cs:19]: Could we move this constant to a better namespace? `BenchmarkDotNet.Toolchains.InProcess.NoEmitBenchmarkActionFactory` looks like a strange place for such an important constant.

Comment (timcassell) [reply src/BenchmarkDotNet/Toolchains/InProcess/NoEmit/BenchmarkActionFactory_Implementations.cs:19]: Do you have a suggestion? I wasn't sure of a best place to put it.

File: src/BenchmarkDotNet/Helpers/Reflection.Emit/MethodBuilderExtensions.cs [hunk 2]
```diff
@@ -37,5 +37,14 @@ public static MethodBuilder SetNoOptimizationImplementationFlag(this MethodBuild
 
             return methodBuilder;
         }
+
+        public static MethodBuilder SetAggressiveOptimizationImplementationFlag(this MethodBuilder methodBuilder)
+        {
+            // AggressiveOptimization is not available in netstandard2.0, so just use the value casted to enum.
+            methodBuilder.SetImplementationFlags(
+                methodBuilder.GetMethodImplementationFlags() | (MethodImplAttributes) 512);
```
[the remark this answers is not in the document]
Comment (timcassell) [reply src/BenchmarkDotNet/Helpers/Reflection.Emit/MethodBuilderExtensions.cs:45]: That's a different enum (`MethodImplOptions` vs `MethodImplAttributes`). I suppose I could make that a constant for better readability anyway.

File: src/BenchmarkDotNet/Toolchains/InProcess/NoEmit/BenchmarkActionFactory_Implementations.cs (hunk 1 above)
Comment (AndreyAkinshin) [reply src/BenchmarkDotNet/Toolchains/InProcess/NoEmit/BenchmarkActionFactory_Implementations.cs:19]: What about `BenchmarkDotNet.Portability`?

File: src/BenchmarkDotNet/Helpers/Reflection.Emit/MethodBuilderExtensions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndreyAkinshin) [reply src/BenchmarkDotNet/Helpers/Reflection.Emit/MethodBuilderExtensions.cs:45]: Yes, please introduce a single constant and put both instances of `MethodImplOptions.AggressiveOptimization` and `MethodImplAttributes.AggressiveOptimization` at the same place.

**Review by AndreyAkinshin — APPROVED** [review]

File: src/BenchmarkDotNet/Portability/CodeGen.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+
+namespace BenchmarkDotNet.Portability
+{
+    public static class CodeGenHelper
+    {
+        // AggressiveOptimization is not available in netstandard2.0, so just use the value casted to enum.
+        public const MethodImplOptions AggressiveOptimizationOption = (MethodImplOptions) 512;
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Portability/CodeGen.cs:9]: is this value ignored by the older runtimes?

cc @AndyAyersMS who added the annotations with ifdefs

Comment (AndyAyersMS) [reply src/BenchmarkDotNet/Portability/CodeGen.cs:9]: Yes, I believe it will be ignored by older .NET Core and all .NET Framework versions.

**Pull request merged**